src/effects/VST/VSTEffect.cpp

```diff
@@ -1128,21 +1128,31 @@ bool VSTEffect::RealtimeResume()
    return true;
 }
 
-bool VSTEffect::RealtimeProcessStart(EffectSettings &)
+bool VSTEffectInstance::RealtimeProcessStart(EffectSettings &)
 {
    return true;
 }
 
-size_t VSTEffect::RealtimeProcess(size_t group, EffectSettings &settings,
+size_t VSTEffectInstance::RealtimeProcess(size_t group, EffectSettings &settings,
    const float *const *inbuf, float *const *outbuf, size_t numSamples)
 {
    wxASSERT(numSamples <= mBlockSize);
    if (group >= mSlaves.size())
       return 0;
-   return mSlaves[group]->ProcessBlock(settings, inbuf, outbuf, numSamples);
+
+   {
+      // If we assume that the user might be moving knobs during realtime processing and wants
+      // to hear how the sound changes, we must do this then
+
+      auto guard = std::lock_guard{ GetEffect().mSettingsMutex };
```

Comment:
Do not do this.

We don't want to do mutex locking and unlocking in a worker thread.

We will trust `settings` to be a copy of settings, properly updated in an atomic fashion, by the `EffectSettingsAccess` object.

Yes, strictly speaking you are right that a data race on the `VSTEffectSettings` stored in the effect, is possible while the transition to statelessness is not complete, and you might remove this mutex locking then.

But I think that is excessive caution.  I didn't transform AudioUnit effects with similar caution and I don't recommend you do it for VST.

Reply:
As I told you during our talk, without introducing this mutex, I always ran into a crash when I moved a slider while processing realtime. This is because mSettings was written in the main thread (when moving the slider), while it was read in the worker thread. 
But I will add a comment about the mutex being a temporary thing until the validator is fully implemented.